src/main/java/io/confluent/connect/jdbc/dialect/DatabaseDialect.java

```diff
@@ -0,0 +1,471 @@
+   *
+   * @param table         the identifier of the table; may not be null
+   * @param keyColumns    the identifiers of the columns in the primary/unique key; may not be null
+   *                      but may be empty
+   * @param nonKeyColumns the identifiers of the other columns in the table; may not be null but may
+   *                      be empty
+   * @return the upsert/merge statement; may not be null
+   */
+  String buildUpsertQueryStatement(
+      TableId table,
+      Collection<ColumnId> keyColumns,
+      Collection<ColumnId> nonKeyColumns
+  );
+
+  /**
+   * Build the DROP TABLE statement expression for the given table.
+   *
+   * @param table  the identifier of the table; may not be null
+   * @param options the options; may be null
+   * @return the DROP TABLE statement; may not be null
+   */
+  String buildDropTableStatement(
+      TableId table,
+      DropOptions options
+  );
+
+  /**
+   * Build the CREATE TABLE statement expression for the given table and its columns.
+   *
+   * @param table  the identifier of the table; may not be null
+   * @param fields the information about the fields in the sink records; may not be null
+   * @return the CREATE TABLE statement; may not be null
+   */
+  String buildCreateTableStatement(
+      TableId table,
+      Collection<SinkRecordField> fields
+  );
+
+  List<String> buildAlterTable(
+      TableId table,
+      Collection<SinkRecordField> fields
+  );
+
+  /**
+   * Create a component that can bind record values into the
+   *
+   * @param statement      the prepared statement
+   * @param pkMode         the primary key mode; may not be null
+   * @param schemaPair     the key and value schemas; may not be null
+   * @param fieldsMetadata the field metadata; may not be null
+   * @param insertMode     the insert mode; may not be null
+   * @return the statement binder; may not be null
+   * @see #bindField(PreparedStatement, int, Schema, Object)
+   */
+  StatementBinder statementBinder(
+      PreparedStatement statement,
+      JdbcSinkConfig.PrimaryKeyMode pkMode,
+      SchemaPair schemaPair,
+      FieldsMetadata fieldsMetadata,
+      JdbcSinkConfig.InsertMode insertMode
+  );
+
+  /**
+   * Method that binds a value with the given schema at the specified variable within a prepared
+   * statement.
+   *
+   * @param statement the prepared statement; may not be null
+   * @param index the 1-based index of the variable within the prepared statement
+   * @param schema the schema for the value; may be null only if the value is null
+   * @param value the value to be bound to the variable; may be null
+   * @throws SQLException if there is a problem binding the value into the statement
+   * @see #statementBinder(PreparedStatement, PrimaryKeyMode, SchemaPair, FieldsMetadata,
+   * InsertMode)
+   */
+  void bindField(
+      PreparedStatement statement,
+      int index,
+      Schema schema,
+      Object value
+  ) throws SQLException;
```

Comment:
This is also something that should be straightforward with the JDBC API, but isn't. Any time a custom dialect overrides the default data type mapping, it may need to change how the values are bound to variables in a prepared statement. For example, `INT8` corresponds to SQL's `BYTE` type, except that SQL `BYTE` is unsigned while `INT8` and Java's `byte` are signed. Most JDBC drivers handle this correctly, but the Sybase driver does nothing special and thus some valid Java `byte` values cannot be inserted into a Sybase `BYTE` column. Therefore, the `SybaseDatabaseDialect` has to instead use `short` to properly handle the signed values, and this changes how `INT8` values are bound to prepared statements for insert/update.